### ISSUE TITLE
v4 Sonarr Sync Fail

### ISSUE THREAD

**Issue by Heavybullets8** [recyclarr/recyclarr#118]
During my testing of Recyclarrs support for sonarr v4. Appreciate the work btw!

OS: Truenas Scale (k3s)
Recyclarr: Docker Edge 
Sonarr: v4 (hotio)

If you need the ENTIRE debug or more than what is shown please let me know.. It was just incredibly long so I only included what I imagine was useful. 

Some of the custom formats actually did get added if i check sonarr, but its seems as if it gets stuck somewhere and ends up just timing out. 

```
[DBG] HTTP Request: POST http://hostname:8989/api/v3/customformat?apikey=SNIP
[DBG] HTTP Body:
{"name":"FreeLeech","includeCustomFormatWhenRenaming":false,"specifications":[{"name":"FreeLeech","implementation":"IndexerFlagSpecification","negate":false,"required":false,"fields":[{"value":1,"name":"value"}]}]}
[DBG] HTTP Response: 500 POST http://hostname:8989/api/v3/customformat?apikey=SNIP
[DBG] HTTP Body:
{
  "message": "Object reference not set to an instance of an object.",
  "description": "System.NullReferenceException: Object reference not set to an instance of an object.\n   at System.Object.GetType()\n   at Sonarr.Api.V3.CustomFormats.CustomFormatResourceMapper.MapSpecification(CustomFormatSpecificationSchema resource, List\u00601 specifications) in C:\\BuildAgent\\work\\13f3e374fa512e16\\src\\Sonarr.Api.V3\\CustomFormats\\CustomFormatResource.cs:line 59\n   at System.Linq.Enumerable.SelectListIterator\u00602.ToList()\n   at Sonarr.Api.V3.CustomFormats.CustomFormatResourceMapper.ToModel(CustomFormatResource resource, List\u00601 specifications) in C:\\BuildAgent\\work\\13f3e374fa512e16\\src\\Sonarr.Api.V3\\CustomFormats\\CustomFormatResource.cs:line 39\n   at Sonarr.Api.V3.CustomFormats.CustomFormatController.Create(CustomFormatResource customFormatResource) in C:\\BuildAgent\\work\\13f3e374fa512e16\\src\\Sonarr.Api.V3\\CustomFormats\\CustomFormatController.cs:line 55\n   at lambda_method176(Closure , Object , Object[] )\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.SyncObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeActionMethodAsync()\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State\u0026 next, Scope\u0026 scope, Object\u0026 state, Boolean\u0026 isCompleted)\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeNextActionFilterAsync()\n--- End of stack trace from previous location ---\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State\u0026 next, Scope\u0026 scope, Object\u0026 state, Boolean\u0026 isCompleted)\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()\n--- End of stack trace from previous location ---\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.\u003CInvokeNextResourceFilter\u003Eg__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State\u0026 next, Scope\u0026 scope, Object\u0026 state, Boolean\u0026 isCompleted)\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()\n--- End of stack trace from previous location ---\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.\u003CInvokeAsync\u003Eg__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.\u003CInvokeAsync\u003Eg__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)\n   at Microsoft.AspNetCore.Routing.EndpointMiddleware.\u003CInvoke\u003Eg__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)\n   at Sonarr.Http.Middleware.BufferingMiddleware.InvokeAsync(HttpContext context) in C:\\BuildAgent\\work\\13f3e374fa512e16\\src\\Sonarr.Http\\Middleware\\BufferingMiddleware.cs:line 28\n   at Sonarr.Http.Middleware.IfModifiedMiddleware.InvokeAsync(HttpContext context) in C:\\BuildAgent\\work\\13f3e374fa512e16\\src\\Sonarr.Http\\Middleware\\IfModifiedMiddleware.cs:line 41\n   at Sonarr.Http.Middleware.CacheHeaderMiddleware.InvokeAsync(HttpContext context) in C:\\BuildAgent\\work\\13f3e374fa512e16\\src\\Sonarr.Http\\Middleware\\CacheHeaderMiddleware.cs:line 33\n   at Sonarr.Http.Middleware.UrlBaseMiddleware.InvokeAsync(HttpContext context) in C:\\BuildAgent\\work\\13f3e374fa512e16\\src\\Sonarr.Http\\Middleware\\UrlBaseMiddleware.cs:line 27\n   at Sonarr.Http.Middleware.VersionMiddleware.InvokeAsync(HttpContext context) in C:\\BuildAgent\\work\\13f3e374fa512e16\\src\\Sonarr.Http\\Middleware\\VersionMiddleware.cs:line 28\n   at Microsoft.AspNetCore.ResponseCompression.ResponseCompressionMiddleware.InvokeCore(HttpContext context)\n   at Microsoft.AspNetCore.Authorization.Policy.AuthorizationMiddlewareResultHandler.HandleAsync(RequestDelegate next, HttpContext context, AuthorizationPolicy policy, PolicyAuthorizationResult authorizeResult)\n   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)\n   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)\n   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.\u003CInvoke\u003Eg__Awaited|6_0(ExceptionHandlerMiddleware middleware, HttpContext context, Task task)"
}
[DBG] Cleaning up logs using max files of 20
HTTP error while communicating with Sonarr: Flurl.Http.FlurlHttpException: Call failed with status code 500 (Internal Server Error): POST http://hostname:8989/api/v3/customformat?apikey=SNIP
   at Flurl.Http.FlurlRequest.HandleExceptionAsync(FlurlCall call, Exception ex, CancellationToken token)
   at Flurl.Http.FlurlRequest.SendAsync(HttpMethod verb, HttpContent content, CancellationToken cancellationToken, HttpCompletionOption completionOption)
   at Flurl.Http.FlurlRequest.SendAsync(HttpMethod verb, HttpContent content, CancellationToken cancellationToken, HttpCompletionOption completionOption)
   at Flurl.Http.ResponseExtensions.ReceiveJson[T](Task`1 response)
   at TrashLib.Services.CustomFormat.Api.CustomFormatService.CreateCustomFormat(ProcessedCustomFormatData cf) in D:\a\recyclarr\recyclarr\src\TrashLib\Services\CustomFormat\Api\CustomFormatService.cs:line 27
   at TrashLib.Services.CustomFormat.Processors.PersistenceSteps.CustomFormatApiPersistenceStep.Process(ICustomFormatService api, CustomFormatTransactionData transactions) in D:\a\recyclarr\recyclarr\src\TrashLib\Services\CustomFormat\Processors\PersistenceSteps\CustomFormatApiPersistenceStep.cs:line 11
   at TrashLib.Services.CustomFormat.Processors.PersistenceProcessor.PersistCustomFormats(IReadOnlyCollection`1 guideCfs, IEnumerable`1 deletedCfsInCache, IDictionary`2 profileScores) in D:\a\recyclarr\recyclarr\src\TrashLib\Services\CustomFormat\Processors\PersistenceProcessor.cs:line 71
   at TrashLib.Services.CustomFormat.CustomFormatUpdater.Process(Boolean isPreview, IEnumerable`1 customFormats, IGuideService guideService) in D:\a\recyclarr\recyclarr\src\TrashLib\Services\CustomFormat\CustomFormatUpdater.cs:line 54
   at Recyclarr.Command.SonarrCommand.Process(IServiceLocatorProxy container) in D:\a\recyclarr\recyclarr\src\Recyclarr\Command\SonarrCommand.cs:line 102
   at Recyclarr.Command.BaseCommand.ExecuteAsync(IConsole console) in D:\a\recyclarr\recyclarr\src\Recyclarr\Command\BaseCommand.cs:line 41
   at Recyclarr.Command.ServiceCommand.ExecuteAsync(IConsole console) in D:\a\recyclarr\recyclarr\src\Recyclarr\Command\ServiceCommand.cs:line 43
```

Config Used for test
```
# Starter config to use with Recyclarr. Most values are set to "reasonable defaults". Update the
# values below as needed for your instance. You will be required to update the API Key and URL for
# each instance you want to use.
#
# Many optional settings have been omitted to keep this template simple.
#
# For more details on the configuration, see the Configuration Reference on the wiki here:
# https://github.com/recyclarr/recyclarr/wiki/Configuration-Reference

# Configuration specific to Sonarr
sonarr:
  # Set the URL/API Key to your actual instance
  - base_url: http://sonarr-v4-custom-app.ix-sonarr-v4.svc.cluster.local:8989
    api_key: ksdjhgksdjfhgskjdghskdjghsdk

    # Quality definitions from the guide to sync to Sonarr. Choice: anime, series, hybrid
    quality_definition: hybrid

    # Release profiles from the guide to sync to Sonarr.
    # You can optionally add tags and make negative scores strictly ignored     
    custom_formats:
      # A list of custom formats to sync to Radarr. Must match the "trash_id" in the guide JSON.
      - trash_ids:
          - b4a1b3d705159cdca36d71e57ca86871 # Anime Raws
#          - 3bc5f395426614e155e585a2f056cdf1 # Season Pack
          - 1656adc6d7bb2c8cca6acfb6592db421 # PCOK
          - 2a7e3be05d3861d6df7171ec74cad727 # PQ
          - 32b367365729d530ca1c124a0b180c64 # Bad Dual Groups
          - f6cce30f1733d5c8194222a7507909bb # HULU
          - 273bd326df95955e1b6c26527d1df89b # v1
          - 63487786a8b01b7f20dd2bc90dd4a477 # DD+
          - a3d82cbef5039f8d295478d28a887159 # HDR10+
          - 06d66ab109d4d2eddb2794d21526d140 # Retags
          - b5a83ef7296f3c5358236e3452ed1d97 # Anime Web Tier 03 (SubsPlease)
          - 29c2a13d091144f63307e4a8ce963a39 # Anime Web Tier 05 (FanSubs)
          - dc262f88d74c651b12e9d90b39f6c753 # Anime Web Tier 06 (FanSubs)
          - 3ac5d84fce98bab1b531393e9c82f467 # QIBI
          - d34870697c9db575f17700212167be23 # NF
          - 7ba05c6e0e14e793538174c679126996 # Multi
          - 0329044e3d9137b08502a9f84a7e58db # Anime BD Tier 07 (P2P/Scene)
          - dbe00161b08a25ac6154c55f95e6318d # DD
          - e3515e519f3b1360cbfc17651944354c # Anime LQ Groups
          - 58790d4e2fdcd9733aa7ae68ba2bb503 # WEB Tier 02
          - bb019e1cd00f304f80971c965de064dc # HDR (undefined)
          - 5964f2a8b3be407d083498e4459d05d0 # DTS
          - e1a997ddb54e3ecbfe06341ad323c458 # Obfuscated
          - 9b64dff695c2115facf1b6ea59c9bd07 # x265 (no HDR/DV)
          - 3497799d29a085e2ac2df9d468413c94 # HDR10
          - 7ea9e9c0fb7cfe4247ad8c915d289cfc # APTV
          - d0c516558625b04b363fa6c5c2c7cfd4 # WEB Scene
          - c429417a57ea8c41d57e6990a8b0033f # DTS-HD MA
          - d84935abd3f8556dcd51d4f27e22d0a6 # WEB Tier 03
          - 0ac24a2a68a9700bcb7eeca8e5cd644c # iT
          - 89358767a60cc28783cdc3d0be9388a4 # DSNP
          - 3e8b714263b26f486972ee1e0fe7606c # MP3
          - c67a75ae4a1715f2bb4d492755ba4195 # PMTP
          - 17e889ce13117940092308f48b48b45b # HMAX
          - bd6dd5e043aa27ff4696a08d011c7d96 # 1.0 Mono
          - b2550eb333d27b75833e25b8c2557b38 # 10bit
          - 949c16fe0a8147f50ba82cc2df9411c9 # Anime BD Tier 01 (Top SeaDex Muxers)
          - cfa5fbd8f02a86fc55d8d223d06a5e1f # DTS-HD HRA
          - 0d7824bb924701997f874e7ff7d4844a # TrueHD ATMOS
          - 4fc15eeb8f2f9a749f918217d4234ad8 # v4
          - 30feba9da3030c5ed1e0f7d610bcadc4 # Anime BD Tier 04 (SeaDex Muxers)
          - e0014372773c8f0e1bef8824f00c7dc4 # Anime Web Tier 01 (Muxers)
          - ec8fa7296b64e8cd390a1600981f3923 # Repack/Proper
          - 418f50b10f1907201b6cfdf881f467b7 # Anime Dual Audio
          - 1895195e84767de180653914ce207245 # 4.0 Sound
          - 44a8ee6403071dd7b8a3a8dd3fe8cb20 # VRV
          - b6fbafa7942952a13e17e2b1152b539a # ATMOS (undefined)
          - 096e406c92baa713da4a72d88030b815 # Anime BD Tier 03 (SeaDex Muxers)
          - 9fb6d778592c293467437593ef394bf1 # 6.1 Surround
          - 3a4127d8aa781b44120d907f2cd62627 # Hybrid
          - 47435ece6b99a0b477caf360e79ba0bb # x265 (HD)
          - d2d7b8a9d39413da5f44054080e028a3 # v0
          - 9b27ab6498ec0f31a3353992e19434ca # DV (WEBDL)
          - 9d00418ba386a083fbf4d58235fc37ef # DTS X
          - 3e2c4e748b64a1a1118e0ea3f4cf6875 # HDR
          - eb3d5cc0a2be0db205fb823640db6a3c # Repack v2
          - 851bd64e04c9374c51102be3dd9ae4cc # FLAC
          - cddfb4e32db826151d97352b8e37c648 # x264
          - e6258996055b9fbab7e9cb2f75819294 # WEB Tier 01
          - 4fd5528a3a8024e6b49f9c67053ea5f3 # Anime Web Tier 04 (Official Subs)
          - 25d2afecab632b1582eaf03b63055f72 # Anime BD Tier 06 (FanSubs)
          - 9c11cd3f07101cdba90a2d81cf0e56b4 # LQ
          - 1f733af03141f068a540eec352589a89 # DV HLG
          - 30f70576671ca933adbdcfc736a69718 # PCM
          - 204c8c3e7315bb0ea81332774fa888d6 # 7.1 Surround
          - ae58039e1319178e6be73caab5c42166 # SHO
          - d7c747094a7c65f4c2de083c24899e8b # FreeLeech
          - 7a235133c87f7da4c8cccceca7e3c7a6 # HBO
          - d6c088c015e1781c4656e36aba94330c # Special Edition
          - 290078c8b266272a5cc8e251b5e2eb0b # 1080p
          - 6d0d8de7b57e35518ac0308b0ddf404e # DV
          - 4232a509ce60c4e208d13825b7c06264 # DD+ ATMOS
          - 7878c33f1963fefb3d6c8657d46c2f0a # DV HDR10
          - ed7f1e315e000aef424a58517fa48727 # Anime BD Tier 02 (SeaDex Muxers)
          - a377864de6228b252d6e28962673cedd # 9.1 Surround
          - 228b8ee9aa0a609463efca874524a6b8 # v2
          - 42cba7e38c7947a6d1d0a62580ee6d62 # 3.0 Sound
          - c1a25cd67b5d2e08287c957b1eb903ec # DTS-ES
          - 85c61753df5da1fb2aab6f2a47426b09 # BR-DISK
          - 3fbafa924f361e66fbc6187af82dfa85 # 5.1 Surround
          - a50b8a0c62274a7c38b09a9619ba9d86 # AAC
          - d660701077794679fd59e8bdf4ce3a29 # AMZN
          - 1bef6c151fa35093015b0bfef18279e5 # 2160p
          - 1b3994c551cbb92a2c781af061f4ab44 # Scene
          - 19180499de5ef2b84b6ec59aae444696 # Anime Web Tier 02 (Top FanSubs)
          - 28f6ef16d61e2d1adfce3156ed8257e3 # Opus
          - 36b72f59f4ea20aad9316f475f2d9fbb # DCU
          - 17e889ce13117940092308f48b48b45b # HLG
          - 4aee45b0868229c4fbd8bad3e315f1d0 # MPEG2
          - 834e534f103938853ffced4203b53e72 # 2.0 Stereo
          - 44e7c4de10ae50265753082e5dc76047 # Repack v3
          - c30d2958827d1867c73318a5a2957eb1 # RED
          - 026d5aadd1a6b4e550b134cb6c72b3ca # Uncensored
          - 0e5833d3af2cc5fa96a0c29cd4477feb # v3
          - c81bbfb47fed3d5a3ad027d077f889de # Anime BD Tier 08 (Mini Encodes)
          - c9eafd50846d299b862ca9bb6ea91950 # x265
          - 27954b0a80aab882522a88a4d9eae1cd # DV SDR
          - 1808e4b9cee74e064dfae3f1db99dbfe # TrueHD
          - b735f09d3c025cbb7d75a5d38325b73b # Remaster
          - 82d40da2bc6923f41e14394075dd4b03 # No-RlsGroup
          - 9c14d194486c4014d422adc64092d794 # Dubs Only
          - 545a76b14ddc349b8b185a6344e28b04 # Anime BD Tier 05 (Remuxes)
        quality_profiles:
          - name: Regular          
```

**Comment by rcdailey** [comment]
What do the logs on the Sonarr side say?

**Comment by rcdailey** [comment]
Ok I did some investigation on this. This is not a Recyclarr issue. The issue is two fold:

- The `docs\json\sonarr\cf\freeleech.json` CF is not valid for Sonarr. It uses `IndexerFlagSpecification` which does not exist. Comparing this to the Radarr version of the same CF, they are identical. I think the moral of the story here is that we cannot just directly copy Radarr CFs to Sonarr.
- The `NullReferenceException` was by itself a defect in my opinion, and I've implemented a solution on the Sonarr side available [here](https://github.com/Sonarr/Sonarr/pull/5146). Note that this does not resolve the error reported here; it just makes it more clear what the error actually is. The required fix is still to export a proper CF compatible with Sonarr.

I'm going to close this issue since there's nothing I can do specifically in Recyclarr to fix this. For now, just avoid using the freeleech CF.
